sam-weekend-lab/public/scripts/article.js

```diff
@@ -0,0 +1,105 @@
+// I want to get the article body and map it (don't change the original), then reduce the number of words in the article to a number (numWordsAll) as a total, using .reduce.  
+Article.numWordsAll = () => {
+  return Article.all.map(request.body.body).reduce(request.body.body)
+};
+
+// Hint: Make sure to return an array and avoid duplicates.
+// I can think that I need to return the array of author names here, but I don't understand how to .reduce them.  Is this asking for a count of all the authors in the table ?  I'm assuming I need to use .filter here somewhere to avoid duplicates as requested.  
+Article.allAuthors = () => {
+  return Article.all.map().reduce();
+};
+
+
+Article.numWordsByAuthor = () => {
+  return Article.allAuthors().map(author => 
+    return {
+      name:
+      // Hint: you will need to chain some combination of .filter(), .map(), and .reduce() to get the value of the numWords property
+      // Pseudocode on this goes like this (I think):
+      // I want to get the number of words per author.  Get the author first.  Find the intersection of all articles published by that author by using a SQL query. Apply the numWordsbyAuthor function to all the articles and get a total by article by using the .map function (don't hange the original array of items.  Take that answer (as an array) and apply the .reduce method to it to add all the numbers of words per articles together to get a sum.
+      numWords:   
+    })
+};
+
+Article.truncateTable = callback => {
+  $.ajax({
+    url: '/articles',
+    method: 'DELETE',
+  })
+    .then(console.log)
+  // REVIEW: Check out this clean syntax for just passing 'assumed' data into a named function! The reason we can do this has to do with the way Promise.prototype.then() works. It's a little outside the scope of 301 material, but feel free to research!
+    .then(callback);
+};
+
+Article.prototype.insertRecord = function(callback) {
+  // REVIEW: Why can't we use an arrow function here for .insertRecord()?
+  // .insertRecord() uses the contextual this.  
+  $.post('/articles', {author: this.author, authorUrl: this.authorUrl, body: this.body, category: this.category, publishedOn: this.publishedOn, title: this.title})
+    .then(console.log)
+    .then(callback);
+};
+
+Article.prototype.deleteRecord = function(callback) {
+  $.ajax({
+    url: `/articles/${this.article_id}`,
+    method: 'DELETE'
+  })
+    .then(console.log)
+    .then(callback);
+};
+
+Article.prototype.updateRecord = function(callback) {
+  $.ajax({
+    url: `/articles/${this.article_id}`,
+    method: 'PUT',
+    data: {
+      author: this.author,
+      authorUrl: this.authorUrl,
+      body: this.body,
+      category: this.category,
+      publishedOn: this.publishedOn,
+      title: this.title,
+      author_id: this.author_id
+    }
+  })
+    .then(console.log)
+    .then(callback);
+};
```

Comment:
module.Article = Article
})(app)

What were doing here is attaching the Article object onto the module object. We pass around this module (also know as app) around our whole application, and because we attach Article and articleView onto module, we have access to all their varibales and functions (app.Article.fetchAll, app.Article.all, app.articleView.populateFilters)